Til and Dominik Review Chapters 4-6+AppC (#157)
scionassociation/scion-cppki_I-D: draft-dekater-scion-pki.md

Reviewer: ```suggestion
- Serial and base number of the current TRC: Including this information allows relying parties to discover TRC updates and trust resets. For specification details, see [](#id).
```
I would use `current` to make clear that the OPTIONAL part is whether to contain the relevant base/serial number. It is NOT OPTIONAL whether the base/serial, if present, are the latest. They must be the latest, or not?

Reviewer 2: Thank you for raising this, I  actually realize that the reference in this paragraph is wrong, because these metadata are actually encoded in a `VerificationKeyID` protobuf message,
so the right reference would be: "AS Entry Signed Header". 

In -controlplane, we already write:

```
   The VerificationKeyID message contains the following REQUIRED fields:

   *  trc_base: Defines the _base_ number of the latest Trust Root
      Configuration (TRC) available to the signer at the time of the
      signature creation.

   *  trc_serial: Defines the _serial_ number of the latest TRC
      available to the signer at the time of the signature creation.
 ```
 
 So I'm wondering if we need to reword the section overall, to align it with what -controlplane says

Reviewer 2: I indeed think we should change the sentence: 
> Additionally, the signer SHOULD include the following information:
 
 And turn it into a MUST, so we align with -dataplane. Asked on [slack](https://scionproto.slack.com/archives/C8ADA9CEP/p1780418195444589)
 
 This SHOULD was introduced in a [very early draft version](https://github.com/scionassociation/scion-cppki_I-D/commit/6db6cc71211767996eac23814968bef0d3368d03) and comes from https://docs.scion.org/en/latest/cryptography/interactions.html#signing-process

Reviewer 2: Continuing this in #164

To verify a received control plane message, the relying party first needs to identify the certificate needed to validate the corresponding signature on the message.

AS certificates are bundled together with the corresponding issuing CA certificate into certificate chains. For efficiency, these certificate chains are distributed separately from the signed messages.

A certificate chain is verified against the control plane root certificate, although the root certificate is bundled with the TRC and **not** in the chain. This makes it possible to extend the validity period of the root certificate and update the corresponding TRC without having to modify the certificate chain.

To verify a control plane message, the relying party MUST perform the following steps:

1. Build a collection of root certificates from the latest TRC of the relevant ISD (that is, the ISD referenced in the signature metadata of the message). If the grace period (see [](#grace)) introduced by the latest TRC is still on-going, the root certificates in the second-to-latest TRC MUST also be included. For a description on how to build the correct collection of certificates, see [](#trc-selection).

> 2. If the signature metadata of the message contains the serial and base number of the latest TRC, the relying party MUST ensure that they have this latest TRC.

Reviewer: ```suggestion
2. The signature metadata of the message contains the serial and base number of the latest TRC. The relying party MUST ensure that they have this latest TRC.
```
If I understand the CP draft correctly, the serial and base are REQUIRED, so there is no "if"...?

https://datatracker.ietf.org/doc/html/draft-dekater-scion-controlplane-18#section-2.2.2.1

Reviewer 2: What if a PKI cert gets used to sign any other types of messages? I think we want to leave PKI as generic as possible, therefore I would leave wording as it is

To verify a received control plane message, the relying party first needs to identify the certificate needed to validate the corresponding signature on the message.

AS certificates are bundled together with the corresponding issuing CA certificate into certificate chains. For efficiency, these certificate chains are distributed separately from the signed messages.

A certificate chain is verified against the control plane root certificate, although the root certificate is bundled with the TRC and **not** in the chain. This makes it possible to extend the validity period of the root certificate and update the corresponding TRC without having to modify the certificate chain.

To verify a control plane message, the relying party MUST perform the following steps:

1. Build a collection of root certificates from the latest TRC of the relevant ISD (that is, the ISD referenced in the signature metadata of the message). If the grace period (see [](#grace)) introduced by the latest TRC is still on-going, the root certificates in the second-to-latest TRC MUST also be included. For a description on how to build the correct collection of certificates, see [](#trc-selection).

> 2. If the signature metadata of the message contains the serial and base number of the latest TRC, the relying party MUST ensure that they have this latest TRC.

Reviewer: I don't quite understand how this relates to signing other types of messages, you mean CP messages other than PCBs? 

I had actually multiple comments on the part, I'll put them here (rephrased and updated):

- It is unclear what "if" means here, maybe it should be clarified: 
  - `if` refers to `contain`, e.g. "if the PCB contains a number" -> this is impossible, because the PCB MUST contain a base and a serial number.
  - `if` refers to `latest`, e.g. "if the contained number is the *latest* number`. 
- How does a receiver know whether this is the latest TRC? I think the real point here is that if an **unknown** TRC number is received, the client MUST request it.
- I think the MUST here can be non-normative. Section 4.2 states: `Consequently, all relying parties MUST be able to verify signatures with the help of the Control Plane PKI.`
  In order to verify a message, the latest certificate must implicitly be available. I 'd say this pretty much implies that the latest TRC is also available?
  - or is the point here that all certificates must be validated with the latest TRC?

I would now rephrase it to something like:

```suggestion
2. If the signature metadata of the message contains the serial and base number of a previously unseen TRC, the relying party MUST ensure that they have this TRC.
```

This is probably still a redundant MUST, and it doesn't say why we need the TRC, but I think it makes it a bit clearer what is meant?

Reviewer 2: Ok, makes senes, I applied your suggestion. Thank you!

It is therefore recommended to deploy multiple, independent CAs within an ISD that can issue certificates to all member ASes and sustain the appropriate certificate renewal load. ASes should then be able to quickly switch over to a backup CA to renew their certificates in time.

> Furthermore, PKI operators need to ensure that the CAs maintain accurate time. Further considerations related to this aspect are discussed in {{I-D.dekater-scion-controlplane}}, section "Effects of Clock Inaccuracy".

Reviewer 2: "Effects of Clock Inaccuracy" only talks about beaconing implications, while "Attacks on Time Sources" explicitly mentions cert. authority. So I'd leave the reference as it was

```suggestion
Furthermore, PKI operators need to ensure that the CAs maintain accurate time. Further considerations related to this aspect are discussed in {{I-D.dekater-scion-controlplane}}, sections "Effects of Clock Inaccuracy" and "Attacks on Time Sources".
```

> Furthermore, PKI operators need to ensure that the CAs maintain accurate time. Further considerations related to this aspect are discussed in {{I-D.dekater-scion-controlplane}}, sections "Effects of Clock Inaccuracy" and "Attacks on Time Sources".

Reviewer 2: ```suggestion
Furthermore, PKI operators need to ensure that the CAs maintain time synchronization with other system components. Further considerations related to this aspect are discussed in {{I-D.dekater-scion-controlplane}}, sections "Effects of Clock Inaccuracy" and "Attacks on Time Sources".
```

As the corresponding PKI messaging only occurs when the control plane is already communicating, these requests to obtain cryptographic material are not prone to additional denial of service attacks. We refer to the security considerations of {{I-D.dekater-scion-controlplane}} for a more detailed description of DoS vulnerabilities of control plane messages.

This does not apply for certificate renewal. Denial of Service on the CA infrastructure or on the communication links from the individual ASes to the CA could be used by an attacker to prevent victim ASes from renewing their certificates and halting the path discovery process. This risk can be mitigated in multiple ways:

- CAs only need to be accessible from ASes within the ISD, reducing the potential DoS attack surface;

> - relying on multiple CAs within an ISD;

Reviewer 2: ```suggestion
- relying on multiple CAs within an ISD (e.g., for certificate renewal);
```

Each representative copies the requested certificates from their machine onto a data exchange device provided by the Ceremony Administrator that is passed between all representatives, before being returned to the Ceremony Administrator. Representatives must not copy the corresponding private keys onto the data exchange device as this invalidates the security of the ceremony.

The Ceremony Administrator then checks that the validity period of each provided certificate covers the previously agreed upon TRC validity, that the signature algorithms are correct, and that the certificate type is valid (root, sensitive voting or regular voting certificate). If these parameters are correct, the Ceremony Administrator computes the SHA-512 hash value for each certificate, aggregates and bundles all the provided certificates, and finally calculates the SHA-512 hash value for the entire bundle. All hash values must be displayed to the participants.

> The Ceremony Administrator musr then share the bundle with the representatives of the Voting ASes who MUST validate on their machine that the hash value of their certificates and that of the bundled certificates is the same as displayed by the Ceremony Administrator.

Reviewer 2: ```suggestion
The Ceremony Administrator must then share the bundle with the representatives of the Voting ASes who MUST validate on their machine that the hash value of their certificates and that of the bundled certificates is the same as displayed by the Ceremony Administrator.
```

The Ceremony Administrator must then share the bundle with the representatives of the Voting ASes who must validate on their machine that the hash value of their certificates and that of the bundled certificates is the same as displayed by the Ceremony Administrator.

This phase concludes when every representative has confirmed the hashes are correct. If there is any mismatch then this phase must be repeated.

### Generation of the TRC Payload {#phase2}

The Ceremony Administrator generates the TRC payload based on the bundled certificates and completed TRC fields (see [](#trcfields)) in accordance with ISD policy, see [](#ceremonyprep).

For each bundled certificate, the voting representatives must then verify the certificate type and that the following fields contain the correct information:

- `issuer`
- `subject`
- `validity`
- `signature`

> Once the voting representatives have verified the TRC data, the Ceremony Administrator computes the DER encoding of the data according to [](#signed-format) and the hash value of the TRC payload file. The TRC payload file is then shared with the voting representatives via the data exchange device who verify the TRC payload hash value by computing it on their machine and checking that it matches the one displayed by the Ceremony Administrator.

Reviewer 2: > and the hash value of the TRC payload file.

This wording becomes a bit vague without mentioning the algorithm. Maybe we restore SHA-512, but say that it could be another one?
